__tests__/commands/upgrade.js

```diff
@@ -102,6 +102,70 @@ test.concurrent('upgrades from fixed version to latest', (): Promise<void> => {
+
+    expect(lockEntryIndex).toEqual(0);
+    expect(lockfile[lockEntryIndex + 1]).toContain('1.1.1');
+    expect(pkg.dependencies).toEqual({'left-pad': '<=1.1.1'});
+  });
+});
+
+test.concurrent('--latest upgrades to latest ignoring package.json when no package name passed', (): Promise<void> => {
+  return runUpgrade([], {latest: true}, 'range-to-latest', async (config): ?Promise<void> => {
+    const lockfile = explodeLockfile(await fs.readFile(path.join(config.cwd, 'yarn.lock')));
+    const pkg = await fs.readJson(path.join(config.cwd, 'package.json'));
+    const lockEntryIndex = lockfile.indexOf('left-pad@^1.1.3:');
+
+    expect(lockEntryIndex).toEqual(0);
+    expect(lockfile.indexOf('left-pad@<=1.1.1:')).toEqual(-1);
+    expect(lockfile[lockEntryIndex + 1]).toContain('1.1.3');
+    expect(pkg.dependencies).toEqual({'left-pad': '^1.1.3'});
+  });
+});
+
+test.concurrent('upgrades to latest matching semver when package name passed with version', (): Promise<void> => {
+  return runUpgrade(['left-pad@~1.1.2'], {}, 'range-to-latest', async (config): ?Promise<void> => {
+    const lockfile = explodeLockfile(await fs.readFile(path.join(config.cwd, 'yarn.lock')));
+    const pkg = await fs.readJson(path.join(config.cwd, 'package.json'));
+    const lockEntryIndex = lockfile.indexOf('left-pad@~1.1.2:');
+
+    expect(lockEntryIndex).toEqual(0);
+    expect(lockfile[lockEntryIndex + 1]).toContain('1.1.3');
+    expect(pkg.dependencies).toEqual({'left-pad': '~1.1.2'});
+  });
+});
+
+test.concurrent('--latest upgrades to passed in version when package name passed with version', (): Promise<void> => {
+  return runUpgrade(['left-pad@1.1.2'], {latest: true}, 'range-to-latest', async (config): ?Promise<void> => {
+    const lockfile = explodeLockfile(await fs.readFile(path.join(config.cwd, 'yarn.lock')));
+    const pkg = await fs.readJson(path.join(config.cwd, 'package.json'));
+    const lockEntryIndex = lockfile.indexOf('left-pad@1.1.2:');
+
+    expect(lockEntryIndex).toEqual(0);
+    expect(lockfile.indexOf('left-pad@<=1.1.1:')).toEqual(-1);
+    expect(lockfile.indexOf('left-pad@^1.1.3:')).toEqual(-1);
+    expect(lockfile[lockEntryIndex + 1]).toContain('1.1.2');
+    expect(pkg.dependencies).toEqual({'left-pad': '1.1.2'});
+  });
+});
+
+test.concurrent('upgrades to latest ignoring package.json semver when package name passed', (): Promise<void> => {
+  return runUpgrade(['left-pad'], {}, 'range-to-latest', async (config): ?Promise<void> => {
+    const lockfile = explodeLockfile(await fs.readFile(path.join(config.cwd, 'yarn.lock')));
+    const pkg = await fs.readJson(path.join(config.cwd, 'package.json'));
+    const lockEntryIndex = lockfile.indexOf('left-pad@^1.1.3:');
+
+    expect(lockEntryIndex).toEqual(0);
+    expect(lockfile.indexOf('left-pad@<=1.1.1:')).toEqual(-1);
+    expect(lockfile[lockEntryIndex + 1]).toContain('1.1.3');
+    expect(pkg.dependencies).toEqual({'left-pad': '^1.1.3'});
+  });
```

Comment:
Yep, you're right. I feel like this should be considered as a bug ... otherwise there's no way to upgrade a specific package inside its allowed boundaries :/ can you open a separate PR where we could discuss this?